streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -303,15 +320,38 @@ private void shutdownTasksAndState(final boolean rethrowExceptions) {
         producer.flush();
         // Close all task state managers
         closeAllStateManagers(rethrowExceptions);
-        try {
-            // un-assign the change log partitions
-            restoreConsumer.assign(Collections.<TopicPartition>emptyList());
-        } catch (Exception e) {
-            log.error("{} Failed to un-assign change log partitions: ", logPrefix, e);
-            if (rethrowExceptions) {
-                throw e;
-            }
-        }
+        // remove the changelog partitions from restore consumer
+        unAssignChangeLogPartitions(rethrowExceptions);
+    }
+
+    /**
+     * Similar to shutdownTasksAndState, however does not close the task managers,
+     * in the hope that soon the tasks will be assigned again
+     * @param rethrowExceptions
+     */
+    private void suspendTasksAndState(final boolean rethrowExceptions) {
+        log.debug("{} suspendTasksAndState: suspending all active tasks [{}] and standby tasks [{}]", logPrefix,
+            activeTasks.keySet(), standbyTasks.keySet());
+
+        // Commit first as there may be cached records that have not been flushed yet.
+        commitOffsets(rethrowExceptions);
+        // flush state
+        flushAllState(rethrowExceptions);
+        // flush out any extra data sent during close
+        producer.flush();
+        // remove the changelog partitions from restore consumer
+        unAssignChangeLogPartitions(rethrowExceptions);
+
+        suspendedTasks.clear();
+        suspendedTasks.putAll(activeTasks);
+        suspendedStandbyTasks.putAll(standbyTasks);
+
+        suspendedTasksByPartition.clear();
+        suspendedTasksByPartition.putAll(activeTasksByPartition);
```

Comment:
We still need to update the `prevTasks` which will be used in the rebalance protocol in order to apply heuristics to reassign tasks to those who previously owned them in best-effort.

So if we do not need `suspendedStandbyTasksByPartition` and `suspendedTasksByPartition` (see above comments), probably we can wrap this logic as `updateSuspendedTasks` along with `removeStreamTasks` and `removeStandbyTasks`, and call all three methods in this case.